### ISSUE TITLE
Function Pointers in ImGuiSettingsHandler are Problematic for Hot-Reloading DLL Systems

### ISSUE THREAD

**Issue by ProfessorSil** [ocornut/imgui#7379]
### Version/Branch of Dear ImGui:

Version 1.40.3 Branch: master

### Back-ends:

Custom Backend (OpenGL based)

### Compiler, OS:

Windows 10, MSVC

### Full config/build information:

```
cl /FeGame.dll /FC /EHsc /EHa- /nologo /GS- /Gm- /GR- /Fm /Oi /WX /W4 /wd4130 /wd4201 /wd4324 /wd4458 /wd4505 /wd4996 /wd4127 /wd4706 /Od /Zi /MTd /wd4065 /wd4100 /wd4101 /wd4127 /wd4189 /wd4702 /I"..\lib\include\imgui" "main.cpp" /link -incremental:no opengl32.lib Ws2_32.lib vulkan-1.lib Dbghelp.lib /EXPORT:Initialize /DLL /PDB:"Game_2024030994249.pdb"
```

### Details:

**Dll Lifespan items in ImGuiContext should be clearly marked and/or be easy to fixup after dll reload**

I've integrated ImGui recently into my game engine which has a dll that holds all the game logic that gets reloaded by a platform layer while maintaining all memory/state information between dll versions. ImGui mostly works with this paradigm as long as I implement allocation functions and repass them after each reload to `ImGui::SetAllocatorFunctions`. This way the context (and everything else) gets allocated from arenas that are sourced from the platform .exe's memory so they are not limited to the DLL lifetime.

The only problems I ran into were with the `ImGuiSettingsHandler` struct which holds function pointers, and string constants, which are both invalid after the DLL reload. I've taken to solving this by doing a post-reload fixup on the Handlers like so:
```
for (ImGuiSettingsHandler& handler : imguiContext->SettingsHandlers)
{
	if (handler.TypeHash == ImHashStr("Window"))
	{
		handler.TypeName = "Window";
		handler.ClearAllFn = WindowSettingsHandler_ClearAll;
		handler.ReadOpenFn = WindowSettingsHandler_ReadOpen;
		handler.ReadLineFn = WindowSettingsHandler_ReadLine;
		handler.ApplyAllFn = WindowSettingsHandler_ApplyAll;
		handler.WriteAllFn = WindowSettingsHandler_WriteAll;
	}
	else if (handler.TypeHash == ImHashStr("Table"))
	{
		handler.TypeName = "Table";
		handler.ClearAllFn = TableSettingsHandler_ClearAll;
		handler.ReadOpenFn = TableSettingsHandler_ReadOpen;
		handler.ReadLineFn = TableSettingsHandler_ReadLine;
		handler.ApplyAllFn = TableSettingsHandler_ApplyAll;
		handler.WriteAllFn = TableSettingsHandler_WriteAll;
	}
	else
	{
		AssertMsg(false, "An ImGuiSettingsHandler was registered for an unknown type. We need to perform proper function pointer fixup for this type on reload, but we don't know how!");
	}
}
```

This solves the problem but it's a bit fragile, in that I have to replicate the code that fills these function pointers in my own code, and they could easily go out of date as I move to newer versions of Dear ImGui.

**Proposal:**
One of the following
1. Add a function like `ImGui::FixupFunctionPointersAndStringConstants(ImGuiContext* context)` which does the fixup like I am doing above. This way the burden to fixup function pointers is put on ImGui and can be maintained appropriately as new function pointers are stored, or function names change.
2. Clearly mark all function pointers and string constants with comments indicating that they will need to be fixed up if the application is going to unload/reload the DLL where they were contextual to. This way it is easy to update the fixup code myself whenever I update to a new version of Dear Imgui. It also makes it clear to other users who have a similar DLL reloading scenario that these bits of data cannot be safely used without fixup when they stick around longer than the DLL they originally came from.
3. Update `ImGuiSettingsHandler` to not require function pointers. There are multiple ways to do this, each with trade-offs. For example, adding an enum that differentiates each type of settings handler and a single function that gets called that can switch off the enum value in the handler and calls the appropriately specialized function. The tradeoff here is it's hard for the application to add their own handlers, so maybe a hybrid approach would be good, where builtin handlers are enum distinguished, and custom handlers have function pointers (that way the application can choose to engage in the function pointer storage if their scenario does not do DLL reloading)

### Minimal, Complete and Verifiable Example code:

```cpp
// I can try and provide a scenario if the description above is unclear. But it would take a decent amount of code (two compilation units, with a loading/reloading mechanism)
```

**Comment by PiggybankStudios** [comment]
Update: `SetPlatformImeDataFn`, `SetClipboardTextFn` and `GetClipboardTextFn` inside `ImGuiIO` are a few more function pointers that need to get updated post dll-reload, since they have DefaultImpl functions they get assigned to, and the application may not realize they need to get updated.

**Comment by ProfessorSil** [comment]
Update: `SetPlatformImeDataFn`, `SetClipboardTextFn` and `GetClipboardTextFn` inside `ImGuiIO` are a few more function pointers that need to get updated post dll-reload, since they have DefaultImpl functions they get assigned to, and the application may not realize they need to get updated.

**Comment by ocornut** [comment]
Yes this would apply to any function pointers, and presumably many more if you use multi-viewports in the docking branch.
I think it may be easier for you to shutdown dear imgui and reinitialize it perhaps, or use an hot-reload system that doesn't invalidate existing pointed-to functions and data that haven't changed. 
While I understand why you would reload game logic/code, I am not even sure I understand what you are aiming to reload dear imgui itself.

**Comment by ProfessorSil** [comment]
To be clear, I don't expect imgui code to change at all from one dll to the next, the goal is not really to reload imgui, it mostly comes as a side effect of the way I have things set up. If I wanted to push imgui into the platform layer I would need to do a decent amount of piping to make it easily accessible by the application dll. It's much easier to just put imgui in the application layer and let it get reloaded along with the application code.

Creating a new context after reload and copying it's pointers is a rather smart idea, I might do that. It doesn't fully solve the problem, it would still be nice if I could verify I've found all function pointers and string constants (i.e. comments clearly marking things would still be helpful).

I realize this might be a somewhat niche use case for imgui, so I'm not surprised it's not fully supported. But I also think it's not far away from being supported (at least, for the master branch, sounds like docking branch would be more work?). I also could do things another way, but this approach I've taken (fixing up function pointers myself) seems to be the most efficient way to get what I want working.

**Comment by ocornut** [comment]
I am not sure what amount of piping you’d need to use but it seems like a more decent solution.
Otherwise, put dear imgui its in own DLL and don’t reload that one, but it’d be faster to not use a DLL for it as it tends to be a call-count-heavy API.

Your suggested solution (1) would assume that nothing has changed or been edited, and thats incorrect imho.
(2) is a burden for us and i’m not convinced is worth it considering i think you perhaps shouldn’t aim to reload or not using this reload tech.
(3) is a constraint for third party extensions indeed.

**Comment by DaedieGit** [comment]
I'm not sure if this is still an issue, but since the issue is still open, I'll chime in with some more detail here for anyone who's interested, since I have a lot of experience with this kind of system.

The correct way to architect a hot-reloading engine is something like this:
application.exe --> engine.dll <-- (reloadable) gamelogic.dll

your engine should hold all state and logic that needs to be persistent and shared between the main application and the reloadable binary.

If you need to incorporate a third party dependency that needs to be exposed to the reloadable, there are 2 ways to do this reasonably:
- add the dependency as a DLL and have all involved parties (engine, gamelogic) link against it.
- have engine.dll integrate it statically, and export the symbols from there. gamelogic doesn't link against the third party directly, but it will receive the symbols through engine.lib. (you could have something like an export.h that includes imgui.h and defines IMGUI_API)

In both cases, you should never hot-reload them. Changes to a third party dependency should require a restart, as they are rarely built to be reloadable. It's also perfectly reasonably to expect this of your users as it should be a very rare thing anyway. (Strictly speaking, you can make them reloadable, but it's considerably more difficult and not relevant in this use case)

With regards to imgui DLL performance. I recently transitioned from static imgui + setting context pointer to imgui.dll and I saw no discernible difference in performance (multi-viewport, average frametime around 0.5ms with imgui currently being the biggest use of cpu time). 
I doubt there is actually that much of a difference on a binary level. You lose out on a bit of linker optimization on the boundary, which isn't likely matter in the grand scheme of things (and usually the kind of sacrifice you're already willing to make to support hot-reloading anyway).
